arraycontext/impl/pytato.py

```diff
@@ -0,0 +1,536 @@
+"""
+.. currentmodule:: arraycontext
+.. autoclass:: PytatoArrayContext
+"""
+__copyright__ = """
+Copyright (C) 2020-1 University of Illinois Board of Trustees
+"""
+
+__license__ = """
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+
+from arraycontext.fake_numpy import \
+        BaseFakeNumpyNamespace, BaseFakeNumpyLinalgNamespace
+from arraycontext.context import ArrayContext
+from arraycontext.container.traversal import \
+        rec_multimap_array_container, rec_map_array_container
+import numpy as np
+from typing import Any, Callable, Tuple, Union, Sequence, Mapping
+from pytools.tag import Tag
+import loopy as lp
+from dataclasses import dataclass, field
+from pyrsistent import pmap, PMap
+
+
+class _PytatoFakeNumpyLinalgNamespace(BaseFakeNumpyLinalgNamespace):
+    # Everything is implemented in the base class for now.
+    pass
+
+
+class _PytatoFakeNumpyNamespace(BaseFakeNumpyNamespace):
+    """
+    A :mod:`numpy` mimic for :class:`PytatoArrayContext`.
+
+
+    .. note::
+
+        :mod:`pytato` does not define any memory layout. If the caller invokes
+        any routine that depends on the input arrays' memory layout, a result is
+        returned assuming the arrays have a C-contiguous layout.
+    """
+    def _get_fake_numpy_linalg_namespace(self):
+        return _PytatoFakeNumpyLinalgNamespace(self._array_context)
+
+    def __getattr__(self, name):
+
+        pt_funcs = ["abs", "sin", "cos", "tan", "arcsin", "arccos", "arctan",
+                    "sinh", "cosh", "tanh", "exp", "log", "log10", "isnan",
+                    "sqrt", "exp"]
+        if name in pt_funcs:
+            import pytato as pt  # type: ignore
+            from functools import partial
+            return partial(rec_map_array_container, getattr(pt, name))
+
+        return super().__getattr__(name)
+
+    def reshape(self, a, newshape):
+        import pytato as pt
```

Comment:
Move the fake-numpy bitss to a submodule so you don't need to have keep importing this. (Analogous issue for the pyopencl array context: #39)

Reply:
Done in https://github.com/inducer/arraycontext/pull/14/commits/e334008d19775abf2fe981284aa6436447e0a409